rust/src/storage/file/mod.rs

```diff
@@ -58,26 +65,43 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
+        let tmp_path = create_tmp_file_with_retry(self, path, obj_bytes).await?;
 
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
-        }
+        rename::rename(&tmp_path, path)
+    }
+}
+
+async fn create_tmp_file_with_retry(
+    backend: &FileStorageBackend,
+    path: &str,
+    obj_bytes: &[u8],
+) -> Result<String, StorageError> {
+    let mut i = 0;
 
-        // use `OpenOptions` with `create_new` to create the file handle.
-        // this will force ErrorKind::AlreadyExists if the file already exists.
-        let mut f = fs::OpenOptions::new()
+    while i < 5 {
```

Comment:
Now I'm a little bit concerning regarding this magic number, how do we usually go about that?

Reply:
leave it as is or move into some static const?

Reply:
I think moving it to a const is a good idea. For optimistic concurrency retries in `DeltaTransaction`, we also provide an [options argument](https://github.com/delta-io/delta-rs/blob/715be9a497ec63b9bcd7dbdae7eba58f15e62049/rust/src/delta.rs#L732-L772), but that might be overkill in the case of the file storage backend.